nixos/modules/services/web-apps/lemmy.nix

```diff
@@ -0,0 +1,142 @@
+{ lib, pkgs, config, ... }:
+with lib;
+let
+  cfg = config.services.lemmy;
+  settingsFormat = pkgs.formats.json { };
+in
+{
+  options.services.lemmy = {
+    enable = mkEnableOption "Enable lemmy server";
+
+    config = mkOption {
+      type = types.submodule {
+        freeformType = settingsFormat.type;
+        options.database = {
+          password = mkOption {
+            type = types.str;
+            description = "password to connect to postgres";
+          };
+        };
+        options.hostname = mkOption {
+          type = types.str;
+          default = null;
+          description = ''the domain name of your instance (eg "lemmy.ml")'';
+        };
+        options.port = mkOption {
+          type = types.port;
+          default = 8536;
+          description = "port where lemmy should listen for incoming requests";
+        };
+        # TODO: this is a secret, should not be in nix-store
+        options.jwt_secret = mkOption {
+          type = types.str;
+          description = "json web token for authorization between server and client";
+        };
+        options.federation = {
+          enabled = mkEnableOption "whether to enable activitypub federation.";
+        };
+        options.captcha = {
+          enabled = mkOption {
+            type = types.bool;
+            default = true;
+            description = "Enable Captcha";
+          };
+          difficulty = mkOption {
+            type = types.str; # TODO: Types should be "easy", "medium", "hard". Define an enum.
+            default = "medium";
+            description = "The difficultly of the captcha to solve";
+          };
+        };
+      };
+      default = { };
+      description = "Lemmy configuration";
+    };
+
+  };
+
+  config = lib.mkIf cfg.enable {
+    services.lemmy.settings = {
+      database.user = mkDefault "lemmy";
+      database.host = mkDefault "localhost";
+      database.port = mkDefault 5432;
+      database.database = mkDefault "lemmy";
+      database.pool_size = mkDefault 5;
+
+      bind = mkDefault "127.0.0.1";
+      tls_enabled = mkDefault true;
+      pictrs_url = mkDefault "http://localhost:8080";
+      actor_name_max_length = mkDefault 20;
+
+      rate_limit.message = mkDefault 180;
+      rate_limit.message_per_second = mkDefault 60;
+      rate_limit.post = mkDefault 6;
+      rate_limit.post_per_second = mkDefault 600;
+      rate_limit.register = mkDefault 3;
+      rate_limit.register_per_second = mkDefault 3600;
+      rate_limit.image = mkDefault 6;
+      rate_limit.image_per_second = mkDefault 3600;
+    };
+
+    # TODO: Check if this can't be put under /var/lib
```

Comment:
Yes.
```
systemd.services.lemmy.environment.LEMMY_CONFIG_LOCATION = "/var/lib/lemmy/config/config.hjson
```

https://github.com/LemmyNet/lemmy/blob/21346eb786dc75efaca19e6de8a514fab3d33da8/crates/utils/src/settings/mod.rs#L44

Reply:
I think here the problem was not so much to say where it is, but more how to get it there.
Now that I think about it with a fresh mind, we should probably use ExecPreStart to copy the configuration over to the right location.

Reply:
Then use `jq` or something to substitute secrets.

Reply:
I think this has been addressed, feel free to comment if it could use improvements.